### PR TITLE
Autowire grid definitions

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Autowires all BO migrated grids and replaced their service ids by FQCN and related services.
| Type?             | refacto
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | yes
| How to test?      | All the migrated grids in BO needs to be checked (only open the grid page and if it is not resulting in fatal error that means it is ok)
| Fixed ticket?     | ADR https://github.com/PrestaShop/ADR/blob/master/0003-use-of-autowiring.md (not sure if there is a dedicated issue for this)
| Related PRs       | If theme, autoupgrade or other module change is needed to make this change work, provide a link to related PRs here.
| Sponsor company   | Your company or customer's name goes here (if applicable).

**Deprecations**
all the current BO lists that was defined in grid_definition_factory,yml are now autowired and their service ids are replaced to FQCN -  :warning: **so all the previous service ids are deprecated** :warning: . As well as following services (now you should use the related FQCN instead):
- prestashop.adapter.data_provider.language
- prestashop.core.grid.action.row.accessibility_checker.print_invoice
- prestashop.core.grid.action.row.accessibility_checker.print_delivery_slip
- prestashop.core.grid.action.row.accessibility_checker.delete_order_return_states
- prestashop.core.grid.action.row.accessibility_checker.delete_order_states
- prestashop.core.grid.action.row.accessibility_checker.delete_order_return_states

### CODE REVIEW

File: src/PrestaShopBundle/Resources/config/services/core/grid/grid_definition_factory.yml [hunk 1]
```diff
@@ -1,418 +1,360 @@
...
+  PrestaShop\PrestaShop\Core\Grid\Definition\Factory\ProductGridDefinitionFactory:
+    arguments:
+      $multistoreFeature: '@PrestaShop\PrestaShop\Adapter\Feature\MultistoreFeature'
+
+  PrestaShop\PrestaShop\Core\Grid\Definition\Factory\CustomerThreadGridDefinitionFactory:
     arguments:
-      - '@prestashop.core.form.choice_provider.language_by_id'
+      $shopNameByIdChoiceProvider: '@prestashop.adapter.form.choice_provider.shop_name_by_id'
+
+  PrestaShop\PrestaShop\Core\Grid\Definition\Factory\LogGridDefinitionFactory: ~
```
Comment (FabienPapet) [inline src/PrestaShopBundle/Resources/config/services/core/grid/grid_definition_factory.yml]: I guess it's time to group ? 😄

**Review by FabienPapet — COMMENTED** [review]
Fantastic work, maybe you can group all those new service definitions, and 👍 for me

File: src/Core/Grid/Definition/Factory/CartRuleGridDefinitionFactory.php [hunk 2]
```diff
@@ -66,9 +67,10 @@ final class CartRuleGridDefinitionFactory extends AbstractGridDefinitionFactory
 
     public function __construct(
         HookDispatcherInterface $hookDispatcher,
+        TranslatorInterface $translator,
         string $contextDateFormat
     ) {
-        parent::__construct($hookDispatcher);
+        parent::__construct($hookDispatcher, $translator);
```
Comment (FabienPapet) [inline src/Core/Grid/Definition/Factory/CartRuleGridDefinitionFactory.php:73]: I'm sceptical about adding optional dependencies inside the constructor. As it can sometimes be useless like here for example.
Perhaps adding a translatable trait, would avoid injecting it everywhere. Avoiding the same error we made for 

WDYT ?

File: src/PrestaShopBundle/Resources/config/services/core/grid/grid_definition_factory.yml [hunk 3]
```diff
@@ -1,418 +1,310 @@
...
+
+  PrestaShop\PrestaShop\Core\Grid\Definition\Factory\AbstractGridDefinitionFactory:
     abstract: true
+
+  PrestaShop\PrestaShop\Core\Grid\Definition\Factory\ProductShopsGridDefinitionFactory:
     arguments:
-      - '@prestashop.core.hook.dispatcher'
-    calls:
-      - [ setTranslator, [ '@translator' ] ]
+      $multistoreFeature: '@PrestaShop\PrestaShop\Adapter\Feature\MultistoreFeature'
```
Comment (FabienPapet) [inline src/PrestaShopBundle/Resources/config/services/core/grid/grid_definition_factory.yml:17]: This can't be autowired ?

**Review by FabienPapet — CHANGES_REQUESTED** [review]

File: src/PrestaShopBundle/Resources/config/services/core/grid/accessibility_checker.yml [hunk 4]
```diff
@@ -1,28 +1,39 @@
...
     class: 'PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\DeleteProfileAccessibilityChecker'
     arguments:
       - '@=service("prestashop.adapter.legacy.configuration").get("_PS_ADMIN_PROFILE_")'
 
+  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\PrintInvoiceAccessibilityChecker: ~
+  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\PrintDeliverySlipAccessibilityChecker: ~
+  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\DeleteOrderReturnStatesAccessibilityChecker: ~
+  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\DeleteOrderStatesAccessibilityChecker: ~
+  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\ProductSingleShopAssociatedAccessibilityChecker: ~
+  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\ProductMultipleShopsAssociatedAccessibilityChecker: ~
```
Comment (zuk3975) [inline src/PrestaShopBundle/Resources/config/services/core/grid/accessibility_checker.yml:18]: todo: group

File: src/PrestaShopBundle/Resources/config/services/core/grid/grid_definition_factory.yml (hunk 3 above)
Comment (zuk3975) [reply src/PrestaShopBundle/Resources/config/services/core/grid/grid_definition_factory.yml:17]: no because it is an interface with multiple implementation, and the one needed is still in adapter

File: src/Core/Grid/Definition/Factory/CartRuleGridDefinitionFactory.php (hunk 2 above)
Comment (zuk3975) [reply src/Core/Grid/Definition/Factory/CartRuleGridDefinitionFactory.php:73]: the trans method must be present in AbstractGridDefinition to avoid BC breaks, but as you suggested I tried to get back the previous implementation of TranslatorAwareTrait and calls: "setTranslator", but had no luck, it required to add "calls: setTranslator to every implementation, I couldn't find a way how to go around it :thinking:

File: src/PrestaShopBundle/Resources/config/services/core/grid/accessibility_checker.yml [hunk 5]
```diff
@@ -3,19 +3,13 @@ services:
...
-  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\PrintInvoiceAccessibilityChecker: ~
-  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\PrintDeliverySlipAccessibilityChecker: ~
-  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\DeleteOrderReturnStatesAccessibilityChecker: ~
-  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\DeleteOrderStatesAccessibilityChecker: ~
-  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\ProductSingleShopAssociatedAccessibilityChecker: ~
-  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\ProductMultipleShopsAssociatedAccessibilityChecker: ~
+  # @todo: couldn't find usages of this service. Probably ProfileGridDefinition is missing it in delete actions.
+  PrestaShop\PrestaShop\Core\Grid\Action\Row\AccessibilityChecker\DeleteProfileAccessibilityChecker:
+    arguments:
+      $superAdminProfileId: '@=service("prestashop.adapter.legacy.configuration").get("_PS_ADMIN_PROFILE_")'
```
Comment (zuk3975) [inline src/PrestaShopBundle/Resources/config/services/core/grid/accessibility_checker.yml]: I couldn't find usages of this service. I think it should be used in ProfileGridDefinition delete action but its out of  scope, so I it should be reproduced and fixed in a dedicated PR.

**Review by FabienPapet — APPROVED** [review]

**Comment by zuk3975** [comment]
@FabienPapet you mentioned this Pr might not be needed anymore when you introduce some gridManager to retrieve the grids no? Can you confirm this is still legit? Else we could close it.

**Comment by FabienPapet** [comment]
Depends of the DI for controllers to be merged :) , if so yes, we can close it.

**Comment by zuk3975** [comment]
this one? https://github.com/PrestaShop/PrestaShop/pull/32450

**Comment by FabienPapet** [comment]
yes

**Comment by zuk3975** [comment]
PR https://github.com/PrestaShop/PrestaShop/pull/32450 got accepted and merged, so this one can be closed
